MapQueryEngineImpl_queryLocalPartition_resultSizeLimitTest.checkResultSize_limitExceeded &  checkResultSize_limitNotExceeded

```
java.lang.AssertionError: Expected exception: com.hazelcast.map.QueryResultSizeExceededException
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:32)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.lang.Thread.run(Thread.java:745)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/1116/testReport/junit/com.hazelcast.map.impl.query/MapQueryEngineImpl_queryLocalPartition_resultSizeLimitTest/checkResultSize_limitExceeded/

```
This exception has been thrown to prevent an OOME on this Hazelcast instance. An OOME might occur when a query collects large data sets from the whole cluster, e.g. by calling IMap.values(), IMap.keySet() or IMap.entrySet(). See GroupProperty.QUERY_RESULT_SIZE_LIMIT for further details. The configured query result size limit is 104223 items. Result size exceeded in local pre-check.
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/1116/testReport/junit/com.hazelcast.map.impl.query/MapQueryEngineImpl_queryLocalPartitions_resultSizeLimitTest/checkResultSize_limitNotExceeded/

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java
@@ -206,7 +206,7 @@ public class MapQueryEngineImpl implements MapQueryEngine {
 
     private void checkQueryResultLimiter(String mapName, Predicate predicate) {
         if (predicate == TruePredicate.INSTANCE) {
-            queryResultSizeLimiter.checkMaxResultLimitOnLocalPartitions(mapName);
+            queryResultSizeLimiter.precheckMaxResultLimitOnLocalPartitions(mapName);
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryResultSizeLimiter.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryResultSizeLimiter.java
@@ -59,6 +59,14 @@ public class QueryResultSizeLimiter {
     public static final float MAX_RESULT_LIMIT_FACTOR = 1.15f;
 
     /**
+     * Adds a security margin to the configured result size limit to prevent false positives in pre-check.
+     * Required since the precheck takes 3 partitions by default and their size may vary due to data distr. imbalance.
+     *
+     * @see QueryResultSizeLimiter
+     */
+    public static final float MAX_RESULT_LIMIT_FACTOR_FOR_PRECHECK = 1.25f;
+
+    /**
      * Special value to mark the disabled state.
      */
     static final int DISABLED = -1;
@@ -106,7 +114,7 @@ public class QueryResultSizeLimiter {
         return isQueryResultLimitEnabled ? (long) ceil(resultLimitPerPartition * ownedPartitions) : Long.MAX_VALUE;
     }
 
-    void checkMaxResultLimitOnLocalPartitions(String mapName) {
+    void precheckMaxResultLimitOnLocalPartitions(String mapName) {
         // check if feature is enabled
         if (!isPreCheckEnabled) {
             return;
@@ -127,7 +135,7 @@ public class QueryResultSizeLimiter {
 
         // check local result size
         long localResultLimit = getNodeResultLimit(partitionsToCheck);
-        if (localPartitionSize > localResultLimit) {
+        if (localPartitionSize > localResultLimit * MAX_RESULT_LIMIT_FACTOR_FOR_PRECHECK) {
             throw new QueryResultSizeExceededException(maxResultLimit, " Result size exceeded in local pre-check.");
         }
     }

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryResultSizeLimiterTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryResultSizeLimiterTest.java
@@ -120,14 +120,14 @@ public class QueryResultSizeLimiterTest {
     public void testLocalPreCheckDisabled() {
         initMocksWithConfiguration(200000, QueryResultSizeLimiter.DISABLED);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test
     public void testLocalPreCheckEnabledWithNoLocalPartitions() {
         initMocksWithConfiguration(200000, 1);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test
@@ -136,7 +136,7 @@ public class QueryResultSizeLimiterTest {
         populatePartitions(partitionsSizes);
         initMocksWithConfiguration(200000, 1);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test
@@ -145,16 +145,16 @@ public class QueryResultSizeLimiterTest {
         populatePartitions(partitionsSizes);
         initMocksWithConfiguration(200000, 1);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test(expected = QueryResultSizeExceededException.class)
     public void testLocalPreCheckEnabledWitPartitionOverLimit() {
-        int[] partitionsSizes = {850};
+        int[] partitionsSizes = {1090};
         populatePartitions(partitionsSizes);
         initMocksWithConfiguration(200000, 1);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test
@@ -163,16 +163,16 @@ public class QueryResultSizeLimiterTest {
         populatePartitions(partitionsSizes);
         initMocksWithConfiguration(200000, 2);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test(expected = QueryResultSizeExceededException.class)
     public void testLocalPreCheckEnabledWitTwoPartitionsOverLimit() {
-        int[] partitionsSizes = {849, 850};
+        int[] partitionsSizes = {1062, 1063};
         populatePartitions(partitionsSizes);
         initMocksWithConfiguration(200000, 2);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test
@@ -181,16 +181,16 @@ public class QueryResultSizeLimiterTest {
         populatePartitions(partitionSizes);
         initMocksWithConfiguration(200000, 2);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test(expected = QueryResultSizeExceededException.class)
     public void testLocalPreCheckEnabledWitMorePartitionsThanPreCheckThresholdOverLimit() {
-        int[] partitionSizes = {849, 850, Integer.MIN_VALUE};
+        int[] partitionSizes = {1200, 1000, Integer.MIN_VALUE};
         populatePartitions(partitionSizes);
         initMocksWithConfiguration(200000, 2);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test
@@ -199,16 +199,16 @@ public class QueryResultSizeLimiterTest {
         populatePartitions(partitionSizes);
         initMocksWithConfiguration(200000, 2);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     @Test(expected = QueryResultSizeExceededException.class)
     public void testLocalPreCheckEnabledWitDifferentPartitionSizesOverLimit() {
-        int[] partitionSizes = {0, 1699, Integer.MIN_VALUE};
+        int[] partitionSizes = {0, 2200, Integer.MIN_VALUE};
         populatePartitions(partitionSizes);
         initMocksWithConfiguration(200000, 2);
 
-        limiter.checkMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
+        limiter.precheckMaxResultLimitOnLocalPartitions(ANY_MAP_NAME);
     }
 
     private void initMocksWithConfiguration(int maxResultSizeLimit, int maxLocalPartitionLimitForPreCheck) {

